File: libraries/cyclestreets-view/src/main/java/net/cyclestreets/views/overlay/RouteHighlightOverlay.java
package net.cyclestreets.views.overlay;

import net.cyclestreets.util.Theme;
import net.cyclestreets.view.R;
import net.cyclestreets.routing.Route;
import net.cyclestreets.routing.Segment;
import net.cyclestreets.views.CycleMapView;

import org.osmdroid.views.MapView;
import org.osmdroid.views.overlay.Overlay;

import android.content.Context;
import android.graphics.Canvas;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;

public class RouteHighlightOverlay extends Overlay
{
  private final CycleMapView mapView;

  private Segment current;

  private final Button routingInfoRect;
  private final ImageView routeNowIcon;
  private final FloatingActionButton prevButton;
  private final FloatingActionButton nextButton;

  private int highlightColour;

  public RouteHighlightOverlay(final Context context, final CycleMapView map) {
    super();

    mapView = map;
    current = null;

    View routeView = LayoutInflater.from(mapView.getContext()).inflate(R.layout.route_view, null);

    routingInfoRect = routeView.findViewById(R.id.routing_info_rect);
    routeNowIcon = routeView.findViewById(R.id.route_now_icon);

    prevButton = routeView.findViewById(R.id.route_highlight_prev);
    prevButton.setVisibility(View.INVISIBLE);
    prevButton.setOnClickListener(view -> regressActiveSegment(1));
    prevButton.setOnLongClickListener(view -> regressActiveSegment(Integer.MAX_VALUE));

    nextButton = routeView.findViewById(R.id.route_highlight_next);
    nextButton.setVisibility(View.INVISIBLE);
    nextButton.setOnClickListener(view -> advanceActiveSegment(1));
    nextButton.setOnLongClickListener(view -> advanceActiveSegment(Integer.MAX_VALUE));

    mapView.addView(routeView);

    highlightColour = Theme.highlightColor(context) | 0xFF000000;
  }

  @Override
  public void draw(final Canvas canvas, final MapView mapView, final boolean shadow) {
    drawButtons();

    if (current == Route.journey().activeSegment())
      return;

    current = Route.journey().activeSegment();
    if (current == null)
      return;

    drawSegmentInfo();
    this.mapView.getController().animateTo(current.start());
  }

  private void drawButtons() {
    if (!Route.available()) {
      prevButton.hide();
      nextButton.hide();
      return;
    }

    prevButton.setEnabled(!Route.journey().atStart());
    prevButton.show();
    nextButton.setEnabled(!Route.journey().atEnd());
    nextButton.show();
  }

  private void drawSegmentInfo() {
    final Segment seg = Route.journey().activeSegment();
    if (seg == null) {
      // In this case, populating the routing info is done by the TapToRouteOverlay
      return;
    }

    routeNowIcon.setVisibility(View.INVISIBLE);

    routingInfoRect.setBackgroundColor(highlightColour);
    routingInfoRect.setGravity(Gravity.LEFT);
    routingInfoRect.setText(seg.toString());
    routingInfoRect.setEnabled(false);
  }

  private boolean regressActiveSegment(int stepsToMove) {
    if (!Route.available()) {
      return false;
    }

    for (int i = stepsToMove; i > 0 && !Route.journey().atStart(); i--)
      Route.journey().regressActiveSegment();

    mapView.invalidate();
    return true;
  }

  private boolean advanceActiveSegment(int stepsToMove) {
    if (!Route.available())
      return false;

    for (int i = stepsToMove; i > 0 && !Route.journey().atEnd(); i--)
      Route.journey().advanceActiveSegment();
    mapView.invalidate();
    return true;
  }

}

File: libraries/cyclestreets-view/src/main/java/net/cyclestreets/views/overlay/RouteHighlightOverlay.kt
package net.cyclestreets.views.overlay

import android.content.Context
import android.graphics.Canvas
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.mikepenz.iconics.typeface.library.googlematerial.GoogleMaterial
import net.cyclestreets.iconics.IconicsHelper.materialIcons
import net.cyclestreets.routing.Journey
import net.cyclestreets.routing.Route.altJourney
import net.cyclestreets.routing.Route.journey
import net.cyclestreets.routing.Route.routeAvailable
import net.cyclestreets.routing.Segment
import net.cyclestreets.util.Theme.lowlightColor
import net.cyclestreets.view.R
import net.cyclestreets.views.CycleMapView
import org.osmdroid.views.MapView
import org.osmdroid.views.overlay.Overlay


class RouteHighlightOverlay(context: Context, private val mapView: CycleMapView) : Overlay() {
// This class highlights the current section of route and displays the turn instruction at the top,
// either in LiveRide, or when the user is pressing the left and right arrow buttons to go through the route.
// Alternatively, it shows the route summary at the top of the screen.

    private var current: Segment? = null
    private var currentAlt: Journey? = null

    private val routeSummaryInfo: TextView
    private val routeNowIcon: ImageView
    private val prevButton: FloatingActionButton
    private val nextButton: FloatingActionButton

    private val altText = context.getString(R.string.alt_route_details)

    init {
        val routeView = LayoutInflater.from(mapView.context).inflate(R.layout.route_view, null)

        routeSummaryInfo = routeView.findViewById(R.id.route_summary_info)
        routeSummaryInfo.visibility = View.GONE
        routeNowIcon = routeView.findViewById(R.id.route_now_icon)

        val (prevIcon, nextIcon) = materialIcons(context, listOf(GoogleMaterial.Icon.gmd_chevron_left, GoogleMaterial.Icon.gmd_chevron_right), lowlightColor(context))

        prevButton = routeView.findViewById<FloatingActionButton>(R.id.route_highlight_prev).apply {
            setImageDrawable(prevIcon)
            visibility = View.INVISIBLE
            setOnClickListener { _ -> regressActiveSegment(1) }
            setOnLongClickListener { _ -> regressActiveSegment(Int.MAX_VALUE) }
        }

        nextButton = routeView.findViewById<FloatingActionButton>(R.id.route_highlight_next).apply {
            setImageDrawable(nextIcon)
            visibility = View.INVISIBLE
            setOnClickListener { _ -> advanceActiveSegment(1) }
            setOnLongClickListener { _ -> advanceActiveSegment(Int.MAX_VALUE) }
        }

        mapView.addView(routeView)

    }

    override fun draw(canvas: Canvas, mapView: MapView, shadow: Boolean) {
        drawButtons()

        if ((current === journey().activeSegment()) && (currentAlt === altJourney()))
            return

        currentAlt = altJourney()
        current = journey().activeSegment()

        drawSegmentInfo()
        if (current != null)
            this.mapView.controller.animateTo(current!!.start())
    }

    private fun drawButtons() {
        if (!routeAvailable()) {
            prevButton.hide()
            nextButton.hide()
            return
        }

        prevButton.isEnabled = !journey().atStart()
        prevButton.show()
        nextButton.isEnabled = !journey().atEnd()
        nextButton.show()
    }

    private fun drawSegmentInfo() {
        val seg = journey().activeSegment()
        val altJourney = altJourney()
        val summaryText: String
        // If there's no active segment, hide the box
        if (seg == null) {
            routeSummaryInfo.text = ""
            // Visibility = GONE means view will not take any space, so button below it will take its place without any gap
            routeSummaryInfo.visibility = View.GONE
            return
        }

        routeNowIcon.visibility = View.INVISIBLE

        // Checking if segment is a Start saves having to define dummy summaryText methods in the other Segment classes
        if ((altJourney == Journey.NULL_JOURNEY) || (seg !is Segment.Start)) {
            summaryText = seg.toString()
        }
        else {
            // The active segment is the Start segment and there is an alt journey.
            summaryText = altJourney?.totalDistance()?.let { seg.summaryText(it, altJourney?.totalTime(), altText) }
                .toString()
        }

        routeSummaryInfo.apply {
            gravity = Gravity.CENTER
            text = summaryText
            visibility = View.VISIBLE
        }
    }

    private fun regressActiveSegment(stepsToMove: Int): Boolean {
        if (!routeAvailable())
            return false

        for (i in stepsToMove downTo 1) {
            if (journey().atStart())
                break
            journey().regressActiveSegment()
        }

        mapView.invalidate()
        return true
    }

    private fun advanceActiveSegment(stepsToMove: Int): Boolean {
        if (!routeAvailable())
            return false

        for (i in stepsToMove downTo 1) {
            if (journey().atEnd())
                break
            journey().advanceActiveSegment()
        }

        mapView.invalidate()
        return true
    }

}